browser/ui/brave_actions/brave_action_icon_with_badge_image_source.cc

```diff
@@ -0,0 +1,98 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/brave_actions/brave_action_icon_with_badge_image_source.h"
+
+#include "base/strings/utf_string_conversions.h"
+#include "cc/paint/paint_flags.h"
+#include "chrome/browser/extensions/extension_action.h"
+#include "chrome/browser/ui/toolbar/toolbar_actions_bar.h"
+#include "chrome/grit/theme_resources.h"
+#include "third_party/skia/include/core/SkColor.h"
+#include "ui/base/resource/resource_bundle.h"
+#include "ui/gfx/canvas.h"
+#include "ui/gfx/color_palette.h"
+#include "ui/gfx/font.h"
+#include "ui/gfx/geometry/rect.h"
+#include "ui/gfx/geometry/size.h"
+#include "ui/gfx/image/image_skia_operations.h"
+#include "ui/gfx/skia_paint_util.h"
+
+void BraveActionIconWithBadgeImageSource::PaintBadge(gfx::Canvas* canvas) {
+    if (!badge_ || badge_->text.empty())
+    return;
+
+  SkColor text_color = SkColorGetA(badge_->text_color) == SK_AlphaTRANSPARENT
+                           ? SK_ColorWHITE
+                           : badge_->text_color;
+
+  SkColor background_color = SkColorSetA(badge_->background_color, SK_AlphaOPAQUE);
+
+  constexpr int kBadgeHeight = 12;
+  constexpr int kPadding = 2;
+  const int kTextHeight = 2 + kBadgeHeight - (kPadding * 2);
```

Comment:
The numbers were generally rendering smaller than the height so I added some ability to bleed in to padding area. But it would be cleaner to just reduce padding on height to `1`, so thanks.